pageserver/src/virtual_file.rs

```diff
@@ -1083,6 +1084,23 @@ impl Drop for VirtualFile {
     }
 }
 
+impl OwnedAsyncWriter for VirtualFile {
+    #[inline(always)]
+    async fn write_all<B: BoundedBuf<Buf = Buf>, Buf: IoBuf + Send>(
+        &mut self,
+        buf: B,
+    ) -> std::io::Result<(usize, B::Buf)> {
+        let (buf, res) = VirtualFile::write_all(self, buf).await;
+        res.map(move |v| (v, buf))
+    }
+
+    #[inline(always)]
+    async fn write_all_borrowed(&mut self, _buf: &[u8]) -> std::io::Result<usize> {
+        // TODO: ensure this through the type system
```

Comment:
Could you expand the commit: it's not obvious why this should be illegal: why can't one write from a &[u8] into a VirtualFile?   Once done, this can probably lose the TODO prefix unless we have a specific idea for how to do it.

Reply:
>  why can't one write from a &[u8] into a VirtualFile?

Because a `&[u8]` is not owned which is incompatible with Rust's async model & io_uring.
Rust's ownership / borrowing system can't model that the kernel holds a reference to the memory pointed to by `&[u8]` while the io_uring operation is ongoing.
Hence Rust allows dropping the future while the operation is still ongoing, which then leads to memory corruption.
That's why I converted the whole write (and read!) path to owned IO.

> Once done, this can probably lose the TODO prefix unless we have a specific idea for how to do it.

I think for the purposes of this coment, it's sufficient to state that

> The VirtualFile API requires owned buffers, not slices, because that's what `tokio_epoll_uring` requires.

Perhaps the `write_all_borrowed` should be an extension trait that is closer to the `InMemoryLayer` where we actually need it.